python/paddle/distributed/auto_parallel/context.py

```diff
@@ -0,0 +1,376 @@
+
+            if node.is_op() and node.op() is not None:
+                op_desc = node.op()
+                op_id = op_desc.id()
+                op_dist_attr = self._op_distributed_attr_map_for_program[op_id]
+                assert op_dist_attr is not None, \
+                    "Operator must have a distributed attribute after the initialization for program."
+                new_op_dist_attr = copy.deepcopy(op_dist_attr)
+                self.set_op_distributed_attr_for_graph(node, new_op_dist_attr)
+        self._is_initialized_for_graph = True
+
+    def clear_distributed_attr_for_program(self):
+        self._tensor_distributed_attr_map_for_program.clear()
+        self._op_distributed_attr_map_for_program.clear()
+
+    def clear_distributed_attr_for_graph(self):
+        self._tensor_distributed_attr_map_for_graph.clear()
+        self._op_distributed_attr_map_for_graph.clear()
+
+    def copy_distribute_attr_from_graph_to_program(self, graph, program):
+        assert self._is_initialized_for_program and self._is_initialized_for_graph, \
+            "The distribute attributes must be initialized both in its program and graph"
+        updated_tensors = {}
+        all_nodes = graph.all_nodes()
+        for node in all_nodes:
+            if node.is_var() and node.var() is not None:
+                tensor_desc = node.var()
+                tensor_id = tensor_desc.id()
+                updated = updated_tensors.get(tensor_desc.name(), False)
+                # If a var has multiples var nodes in graph, only use the first one for now
+                if not updated:
+                    tensor_dist_attr = self.get_tensor_distributed_attr_for_graph(
+                        node)
+                    new_tensor_dist_attr = copy.deepcopy(tensor_dist_attr)
+                    self._tensor_distributed_attr_map_for_program[
+                        tensor_id] = new_tensor_dist_attr
+                    updated_tensors[tensor_desc.name()] = True
+            if node.is_op() and node.op() is not None:
+                op_desc = node.op()
+                op_id = op_desc.id()
+                op_dist_attr = self.get_op_distributed_attr_for_graph(node)
+                new_op_dist_attr = copy.deepcopy(op_dist_attr)
+                self._op_distributed_attr_map_for_program[
+                    op_id] = new_op_dist_attr
+
+    def amend_distributed_attr_for_program(self):
+        for attr in self._tensor_distributed_attr_map_for_program.values():
+            assert attr.is_valid(), \
+                "Tensor's distributed attribute {} is not valid".format(attr)
+            tensor_shape = attr.get_shape()
+            dims_mapping = attr.get_dims_mapping()
+            process_mesh_shape = attr.get_process_mesh().topology
+            # If the dimension of tensor is less than the sharding dimension of process mesh,
+            # we just amend the dimension mapping to -1. (Is this really OK?)
```

Comment:
in this case it should raise an error directly?
or 
we need define a un-even sharding rule for it allowing only part of process hold a real shard of the tensor, and part of them hold a null shard ?

Reply:
Good point. This part should be implemented in a better way. But this is not an error, and it's a design choice.